src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java

```diff
@@ -72,152 +80,143 @@ private void addReject(long val) {
         }
     }
 
-    private final Matcher include;
-    private final Matcher exclude;
-    private final CharsRefBuilder scratch = new CharsRefBuilder();
-    private Set<BytesRef> includeValues;
-    private Set<BytesRef> excludeValues;
-    private final boolean hasRegexTest;
+    public static class StringFilter {
+
+        private final ByteRunAutomaton runAutomaton;
+
+        private StringFilter(Automaton automaton) {
+            this.runAutomaton = new ByteRunAutomaton(automaton);
+        }
+
+        /**
+         * Returns whether the given value is accepted based on the {@code include} & {@code exclude} patterns.
+         */
+        public boolean accept(BytesRef value) {
+            return runAutomaton.run(value.bytes, value.offset, value.length);
+        }
+    }
+
+    public static class OrdinalsFilter {
+
+        private final CompiledAutomaton compiled;
+
+        private OrdinalsFilter(Automaton automaton) {
+            this.compiled = new CompiledAutomaton(automaton);
+        }
+
+        /**
+         * Computes which global ordinals are accepted by this IncludeExclude instance.
+         */
+        public LongBitSet acceptedGlobalOrdinals(RandomAccessOrds globalOrdinals, ValuesSource.Bytes.WithOrdinals valueSource) throws IOException {
+            LongBitSet acceptedGlobalOrdinals = new LongBitSet(globalOrdinals.getValueCount());
+            TermsEnum globalTermsEnum;
+            Terms globalTerms = new DocValuesTerms(globalOrdinals);
+            globalTermsEnum = compiled.getTermsEnum(globalTerms);
```

Comment:
Can you add a TODO here? Something like: 

    // TODO: specialize based on compiled.type: for ALL and prefixes (sinkState >= 0 ) we can avoid i/o and just set bits.

I can look into it later.